packages/lib-utils/src/k8s/hooks/useK8sWatchResources.ts

```diff
@@ -0,0 +1,184 @@
+  WatchK8sResults,
+} from './watch-resource-types';
+
+/**
+ * Hook that retrieves the k8s resources along with their respective status for loaded and error.
+ * @param initResources resources need to be watched as key-value pair, wherein key will be unique to resource and value will be options needed to watch for the respective resource.
+ * @return A map where keys are as provided in initResouces and value has three properties data, loaded and error.
+ * @example
+ * ```ts
+ * const Component: React.FC = () => {
+ *   const watchResources = {
+        'deployment': {...},
+        'pod': {...}
+        ...
+      }
+ *   const {deployment, pod} = useK8sWatchResources(watchResources)
+ *   return ...
+ * }
+ * ```
+ */
+export const useK8sWatchResources = <R extends ResourcesObject>(
+  initResources: WatchK8sResources<R>,
+): WatchK8sResults<R> => {
+  const cluster = useSelector((state: SDKStoreState) => getActiveCluster(state));
+  const resources = useDeepCompareMemoize(initResources, true);
+  const modelsLoaded = useModelsLoaded();
+
+  const allK8sModels = useSelector<SDKStoreState, ImmutableMap<string, K8sModelCommon>>(
+    (state: SDKStoreState) => state.k8s?.getIn(['RESOURCES', 'models']),
+  );
+
+  const prevK8sModels = usePrevious(allK8sModels);
+  const prevResources = usePrevious(resources);
+
+  // Ref to immutable map mapping model's kind (string) to the model object (ImmutableMap)
+  const k8sModelsRef = React.useRef<ImmutableMap<string, K8sModelCommon>>(ImmutableMap());
+
+  if (
+    prevResources !== resources ||
+    (prevK8sModels !== allK8sModels &&
+      Object.values(resources).some((r) => {
+        // TODO: Use model in lieu of modelReference
+        const modelReference = transformGroupVersionKindToReference(r.groupVersionKind || r.kind);
+        return prevK8sModels?.get(modelReference) !== allK8sModels?.get(modelReference);
+      }))
+  ) {
+    // String array containing list of model kinds/GVK based on input resources
+    const requiredModels = Object.values(resources).map((r) =>
+      transformGroupVersionKindToReference(r.groupVersionKind || r.kind),
+    );
+
+    // Filter all models to get the specific models required for the watched resources
+    k8sModelsRef.current = allK8sModels.filter((model) => {
+      return model
+        ? requiredModels.includes(getReferenceForModel(model)) ||
+            requiredModels.includes(model.kind)
+        : false;
+    }) as ImmutableMap<string, K8sModelCommon>;
+  }
+
+  // Contains mapping of model's kind/GVK (string) to the specific model objects to be watched (ImmutableMap)
+  const k8sModels = k8sModelsRef.current;
+
+  type WatchModel = ReturnType<GetWatchData> & { noModel: boolean };
+
+  // reduxIDs -- Map of keys from "resources" to the {id, action} for watching the specific resource
+  const reduxIDs = React.useMemo<{
+    [key: string]: WatchModel;
+  } | null>(
+    () =>
+      modelsLoaded
+        ? Object.keys(resources).reduce(
```

Comment:
Same here as above, there is quite a bit of logic nested in a ternary here and it makes readability hard.